jf_agent/bb_download.py

```diff
@@ -8,13 +8,51 @@
 from requests.packages.urllib3.util.retry import Retry
 
 
-def get_session():
+class StashySession(requests.Session):
+    def __init__(self, base_url, username, password, verify):
+        super().__init__()
+
+        self._client = None
+        self.base_url = base_url
+        self.username = username
+        self.password = password
+        self.verify = verify
+
+    def save_client_reference(self, client):
+        self._client = client
+
+    def reset_client_session(self):
+        self._client._client = stashy.client.StashClient(
+            self.base_url,
+            username=self.username,
+            password=self.password,
+            verify=self.verify,
+            session=self,
+        )
+
+    def request(self, method, url, **kwargs):
+        # If we get HTTP 401, re-authenticate and try again
+        response = super().request(method, url, **kwargs)
+        if response.status_code == 401:
+            print(f'WARN: received 401 for the request [{method}] {url} - resetting client session')
+            self._client._client = stashy.client.StashClient(
+                self.base_url,
+                username=self.username,
+                password=self.password,
+                verify=self.verify,
+                session=self,
+            )
```

Comment:
Should this use `self.reset_client_session()`?

Reply:
whoops